### PR TITLE
pally: Add histograms for tracking latency

### PR DESCRIPTION
**Pull request by akshayjshah**
To round out the basic metric types, we need some way to track RPC latencies.
The best approach is to use histograms - it's both more correct and closer to the
stats that Muttley collects.

This PR adds scalar and vectorized histograms to Pally, along with test
coverage. (See the package-level Pally documentation for an explanation of
vector types).

Down the road, we'll need to come back and optimize the Tally export path; right
now, we take the simple approach of just calling
`tally.Histogram.ObserveDuration` multiple times with the same value. This isn't
ideal, since it forces Tally to binary search for the correct histogram bucket
repeatedly.

Reviewers, the details of this code require some understanding of the [Prometheus](https://godoc.org/github.com/prometheus/client_golang/prometheus#Histogram
) and [Tally](https://godoc.org/github.com/uber-go/tally#Histogram) histogram implementations.

This is one sub-task of T841911.

### CODE REVIEW

File: internal/pally/metric.go [hunk 1]
```diff
@@ -66,6 +68,23 @@ type GaugeVector interface {
 	MustGet(...string) Gauge
 }
 
+// Latencies approximates a latency distribution with a histogram.
+//
+// Latencies are exported to both Prometheus and Tally using their native
+// histogram types.
+type Latencies interface {
+	Observe(time.Duration)
```
Comment (willhug) [inline internal/pally/metric.go]: (non-blocking)  Might be nice to have an indication that the durations passed in here should not be changed.  I can see people shooting themselves in the foot by adjusting these values based on their metric "unit"

File: internal/pally/histogram.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/uber-go/tally"
+	"go.uber.org/atomic"
+)
+
+type bucket struct {
+	atomic.Int64
+
+	last  int64
```
Comment (akshayjshah) [inline internal/pally/histogram.go]: √

File: internal/pally/histogram.go [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+	upper int64 // inclusive
```
Comment (akshayjshah) [inline internal/pally/histogram.go]: added a clarifying comment.

File: internal/pally/histogram.go [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+		} else {
+			j = h
+		}
+	}
+	return bs[i]
+}
+
+type histogram struct {
+	buckets buckets
+	sum     atomic.Int64
```
Comment (akshayjshah) [inline internal/pally/histogram.go]: Same, will add a comment. This is the sum for a single histogram, though.

File: internal/pally/histogram.go [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+
+	opts              LatencyOpts
+	desc              *prometheus.Desc
+	tally             tally.Histogram
+	variableLabelVals []string
+}
+
+func newHistogram(opts LatencyOpts) *histogram {
+	return &histogram{
+		buckets: opts.buckets(),
```
Comment (akshayjshah) [inline internal/pally/histogram.go]: Yep, will do in a separate PR. Tracked in phab.

File: internal/pally/histogram.go [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+			h.tally.RecordDuration(time.Duration(bucket.upper) * h.opts.Unit)
+		}
+	}
+}
+
+func (h *histogram) Collect(ch chan<- prometheus.Metric) {
+	// TODO: Implement prometheus.Metric directly, which avoids all these
+	// allocations.
+	n := uint64(0)
+	prom := make(map[float64]uint64, len(h.buckets)-1)
```
Comment (akshayjshah) [inline internal/pally/histogram.go]: Done,

File: internal/pally/histogram.go [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+	if err == nil {
+		ch <- m
+	}
+}
+
+func (h *histogram) Describe(ch chan<- *prometheus.Desc) {
+	ch <- h.desc
+}
+
+type histogramVector struct {
```
Comment (akshayjshah) [inline internal/pally/histogram.go]: Covered in package-level docs and docstrings for exported interfaces.

File: internal/pally/histogram.go [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+func (vec *histogramVector) newHistogram(key []byte, variableLabelVals []string) (*histogram, error) {
+	m, ok := vec.histograms[string(key)]
+	if ok {
+		return m, nil
+	}
+	if len(vec.opts.VariableLabels) != len(variableLabelVals) {
+		return nil, errInconsistentCardinality
+	}
+	m = &histogram{
+		buckets:           vec.bs,
```
Comment (akshayjshah) [inline internal/pally/histogram.go]: Great catch, fixed and added a regression test.

File: internal/pally/histogram.go [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+	ch <- h.desc
+}
+
+type histogramVector struct {
+	opts LatencyOpts
+	desc *prometheus.Desc
+	bs   buckets
+
+	histogramsMu sync.RWMutex
+	histograms   map[string]*histogram
```
Comment (akshayjshah) [inline internal/pally/histogram.go]: done.

File: internal/pally/opts.go [hunk 10]
```diff
@@ -83,3 +85,53 @@ func (o Opts) copyLabels() map[string]string {
 	}
 	return l
 }
+
+// LatencyOpts configure Latencies and LatenciesVectors.
+type LatencyOpts struct {
+	Opts
+
+	Unit    time.Duration
+	Buckets []time.Duration // upper bounds
```
Comment (akshayjshah) [inline internal/pally/opts.go]: Documented somewhat, but this will probably be clarified best by examples down the road.

File: internal/pally/metric.go (hunk 1 above)
Comment (akshayjshah) [inline internal/pally/metric.go]: I think that this should be captured by the type; we're asking for a time.Duration, which should clearly signal that there's no multiplication or division required.

**Comment by akshayjshah** [comment]
@willhug updated, PTAL.
